File: challenges/groceries/bot/coordinator.py
"""
Multi-bot coordinator.

Processes bots in priority order and resolves movement conflicts via a
reservation table. Bots holding deliverable items move first.

Priority score (higher = moves first):
  holding_useful * 1000 + inventory_count * 100 - dist_to_nearest_dropoff
"""
from __future__ import annotations
from typing import Optional

from state import Bot, Item, GameState
from pathfinding import precompute_drop_distances
from assignment import assign_tasks
from planner import plan_bot_action


DIRECTION_DELTAS = {
    "move_up":    (0, -1),
    "move_down":  (0,  1),
    "move_left":  (-1, 0),
    "move_right": (1,  0),
}


def apply_move(pos: tuple[int, int], action: str) -> tuple[int, int]:
    if action in DIRECTION_DELTAS:
        dx, dy = DIRECTION_DELTAS[action]
        return (pos[0] + dx, pos[1] + dy)
    return pos  # pick_up / drop_off / wait — bot stays in place


def priority_score(
    bot: Bot,
    state: GameState,
    drop_dist: dict[tuple[int, int], int],
) -> int:
    active = state.active_order
    active_needed = active.still_needed if active else []
    holding_useful = any(i in active_needed for i in bot.inventory)
    inv_count = len(bot.inventory)
    dist = drop_dist.get(bot.position, 9999)
    return int(holding_useful) * 1000 + inv_count * 100 - dist


def plan_round(
    state: GameState,
    drop_dist: dict[tuple[int, int], int],
) -> list[dict]:
    """
    Plans actions for all bots for this round.
    Returns a list of action dicts suitable for the WebSocket response.
    """
    walls = state.grid.walls
    w, h = state.grid.width, state.grid.height
    drop_zones = state.drop_off_zones

    # Needed items for active order (excluding in-transit)
    active = state.active_order
    truly_needed_types = state.truly_needed(active) if active else []
    available_items = state.available_items_for(truly_needed_types)

    # Task assignment: bot_id → Item | None
    assignment = assign_tasks(state.bots, available_items, state, drop_dist, walls)

    # Sort bots by priority (highest first)
    bots_ordered = sorted(
        state.bots,
        key=lambda b: -priority_score(b, state, drop_dist),
    )

    # Reservation table: next_cell → bot_id
    reserved: dict[tuple[int, int], int] = {}

    # Current positions (to detect swap deadlocks)
    current_positions = {b.id: b.position for b in state.bots}

    # Congestion counter for drop-off zones
    dropoff_congestion: dict[tuple[int, int], int] = {}
    for bot in state.bots:
        if bot.position in drop_zones:
            dropoff_congestion[bot.position] = dropoff_congestion.get(bot.position, 0) + 1

    results: dict[int, dict] = {}

    for bot in bots_ordered:
        pos = bot.position

        # Cells already reserved by higher-priority bots
        blocked = frozenset(reserved.keys()) - {pos}

        action_dict = plan_bot_action(
            bot=bot,
            state=state,
            assigned_item=assignment.get(bot.id),
            drop_dist=drop_dist,
            walls=walls,
            blocked=blocked,
            dropoff_congestion=dropoff_congestion,
        )

        action = action_dict.get("action", "wait")
        desired_cell = apply_move(pos, action)

        # Check for swap deadlock: if desired_cell is occupied by a bot that
        # wants to move to our current cell
        swap_deadlock = False
        for other_id, other_result in results.items():
            other_action = other_result.get("action", "wait")
            other_pos = current_positions[other_id]
            other_desired = apply_move(other_pos, other_action)
            if desired_cell == other_pos and other_desired == pos:
                swap_deadlock = True
                break

        # Resolve conflicts
        if desired_cell in reserved or swap_deadlock:
            # Try perpendicular sidestep
            sidestep = _find_sidestep(pos, action, walls, reserved, w, h)
            if sidestep:
                action = sidestep
                desired_cell = apply_move(pos, sidestep)
                action_dict = {"action": sidestep}
            else:
                # Give up — wait in place
                action = "wait"
                desired_cell = pos
                action_dict = {"action": "wait"}

        reserved[desired_cell] = bot.id
        results[bot.id] = action_dict

    # Build final output list
    output = []
    for bot in state.bots:
        action_dict = results[bot.id]
        entry = {"bot": bot.id}
        entry.update(action_dict)
        output.append(entry)

    return output


def _find_sidestep(
    pos: tuple[int, int],
    intended_action: str,
    walls: frozenset[tuple[int, int]],
    reserved: dict[tuple[int, int], int],
    width: int,
    height: int,
) -> Optional[str]:
    """Try perpendicular moves to the intended direction."""
    perp = {
        "move_up":    ["move_left", "move_right"],
        "move_down":  ["move_left", "move_right"],
        "move_left":  ["move_up", "move_down"],
        "move_right": ["move_up", "move_down"],
    }
    candidates = perp.get(intended_action, [])
    for action in candidates:
        dx, dy = DIRECTION_DELTAS[action]
        nx, ny = pos[0] + dx, pos[1] + dy
        npos = (nx, ny)
        if (
            0 <= nx < width
            and 0 <= ny < height
            and npos not in walls
            and npos not in reserved
        ):
            return action
    return None

File: challenges/groceries/bot/coordinator.ts
import type { Bot, GameState, Position, BotAction, RoundAction, DistanceMap } from "./types.ts";
import { posKey } from "./utils.ts";
import { assignTasks } from "./assignment.ts";
import { planBotAction } from "./planner.ts";
import { stillNeeded, getActiveOrder, trulyNeeded, availableItemsFor } from "./state.ts";

export const DIRECTION_DELTAS: Record<string, [number, number]> = {
  move_up: [0, -1],
  move_down: [0, 1],
  move_left: [-1, 0],
  move_right: [1, 0],
};

export function applyMove(pos: Position, action: string): Position {
  const delta = DIRECTION_DELTAS[action];
  if (delta) {
    return [pos[0] + delta[0], pos[1] + delta[1]];
  }
  return pos; // pick_up / drop_off / wait — bot stays in place
}

export function priorityScore(
  bot: Bot,
  state: GameState,
  dropDist: DistanceMap,
): number {
  const active = getActiveOrder(state);
  const activeNeeded = active ? stillNeeded(active) : [];
  const holdingUseful = bot.inventory.some((item) => activeNeeded.includes(item));
  const invCount = bot.inventory.length;
  const dist = dropDist.get(posKey(bot.position)) ?? 9999;
  return (holdingUseful ? 1000 : 0) + invCount * 100 - dist;
}

export function planRound(
  state: GameState,
  dropDist: DistanceMap,
): RoundAction[] {
  const walls = state.grid.walls;
  const { width, height } = state.grid;

  // Needed items for active order (excluding in-transit)
  const active = getActiveOrder(state);
  const trulyNeededTypes = active ? trulyNeeded(state, active) : [];
  const availableItems = availableItemsFor(state, trulyNeededTypes);

  // Task assignment: bot_id → Item | null
  const assignment = assignTasks(state.bots, availableItems, state, dropDist, walls);

  // Sort bots by priority (highest first)
  const botsOrdered = [...state.bots].sort(
    (a, b) => priorityScore(b, state, dropDist) - priorityScore(a, state, dropDist),
  );

  // Reservation table: posKey → bot_id
  const reserved = new Map<string, number>();

  // Current positions (to detect swap deadlocks)
  const currentPositions = new Map<number, Position>();
  for (const bot of state.bots) {
    currentPositions.set(bot.id, bot.position);
  }

  const results = new Map<number, BotAction>();

  for (const bot of botsOrdered) {
    const pos = bot.position;

    // Cells already reserved by higher-priority bots (exclude own position)
    const blocked = new Set<string>();
    for (const [key] of reserved) {
      if (key !== posKey(pos)) {
        blocked.add(key);
      }
    }

    const actionDict = planBotAction(
      bot,
      state,
      assignment.get(bot.id) ?? null,
      dropDist,
      walls,
      blocked,
    );

    let action = actionDict.action;
    let desiredCell = applyMove(pos, action);

    // Check for swap deadlock
    let swapDeadlock = false;
    for (const [otherId, otherAction] of results) {
      const otherPos = currentPositions.get(otherId)!;
      const otherDesired = applyMove(otherPos, otherAction.action);
      const desiredKey = posKey(desiredCell);
      const otherPosKey = posKey(otherPos);
      const posKeyStr = posKey(pos);
      const otherDesiredKey = posKey(otherDesired);
      if (desiredKey === otherPosKey && otherDesiredKey === posKeyStr) {
        swapDeadlock = true;
        break;
      }
    }

    // Resolve conflicts
    if (reserved.has(posKey(desiredCell)) || swapDeadlock) {
      const sidestep = findSidestep(pos, action, walls, reserved, width, height);
      if (sidestep) {
        action = sidestep;
        desiredCell = applyMove(pos, sidestep);
        results.set(bot.id, { action: sidestep });
      } else {
        action = "wait";
        desiredCell = pos;
        results.set(bot.id, { action: "wait" });
      }
    } else {
      results.set(bot.id, actionDict);
    }

    reserved.set(posKey(desiredCell), bot.id);
  }

  // Build final output list (preserve original bot order)
  const output: RoundAction[] = [];
  for (const bot of state.bots) {
    const actionDict = results.get(bot.id)!;
    const entry: RoundAction = { bot: bot.id, action: actionDict.action };
    if (actionDict.item_id) {
      entry.item_id = actionDict.item_id;
    }
    output.push(entry);
  }

  return output;
}

export function findSidestep(
  pos: Position,
  intendedAction: string,
  walls: Set<string>,
  reserved: Map<string, number>,
  width: number,
  height: number,
): string | null {
  const perp: Record<string, string[]> = {
    move_up: ["move_left", "move_right", "move_down"],
    move_down: ["move_left", "move_right", "move_up"],
    move_left: ["move_up", "move_down", "move_right"],
    move_right: ["move_up", "move_down", "move_left"],
  };

  const candidates = perp[intendedAction] ?? [];
  for (const action of candidates) {
    const [dx, dy] = DIRECTION_DELTAS[action];
    const nx = pos[0] + dx;
    const ny = pos[1] + dy;
    const nKey = posKey([nx, ny]);
    if (
      nx >= 0 &&
      nx < width &&
      ny >= 0 &&
      ny < height &&
      !walls.has(nKey) &&
      !reserved.has(nKey)
    ) {
      return action;
    }
  }
  return null;
}
